src/tree_math.cpp

```diff
@@ -57,9 +57,25 @@ LeafIndex::LeafIndex(NodeIndex x)
   val = x.val >> 1; // NOLINT(hicpp-signed-bitwise)
 }
 
-NodeIndex::NodeIndex(LeafIndex x)
-  : UInt32(2 * x.val)
+NodeIndex
+LeafIndex::ancestor(LeafIndex other) const
 {
+  auto ln = NodeIndex(*this);
+  auto rn = NodeIndex(other);
+  if (ln == rn) {
+    return ln;
+  }
+
+  uint8_t k = 0;
+  while (ln != rn) {
+    ln.val = ln.val >> 1U;
+    rn.val = rn.val >> 1U;
+    k += 1;
+  }
+
+  uint32_t prefix = ln.val << k;
+  uint32_t stop = (1U << uint8_t(k - 1));
+  return NodeIndex{ prefix + (stop - 1) };
```

Comment:
wondering if a basic test would be useful ?

Reply:
Good idea, added ancestor to test vectors.